### ISSUE TITLE
Assertion in C++ triggered by trying to use field of Shared mem vector

### ISSUE THREAD

**Issue by oliver-batchelor** [taichi-dev/taichi#8435]
```
import taichi as ti

@ti.kernel
def foo():

  for x in range(1000):
    shared = ti.simt.block.SharedArray((10, ), dtype=ti.math.vec3)
    print(shared[x].z) # Asserts 
    print(shared[x]) # Also asserts


ti.init(arch=ti.cuda)
foo()

```

```
[Taichi] version 1.7.0, llvm 15.0.4, commit 59156b3c, linux, python 3.10.11
[Taichi] Starting on arch=cuda
[E 12/13/23 20:10:29.261 521918] [frontend_ir.cpp:is_global@850] Assertion failure: var.cast<IndexExpression>()->is_matrix_field() || var.cast<IndexExpression>()->is_ndarray()


Traceback (most recent call last):
  File "/home/oliver/sync/bad_shared.py", line 13, in <module>
    foo()
  File "/home/oliver/mambaforge/envs/torch2/lib/python3.10/site-packages/taichi/lang/kernel_impl.py", line 1103, in wrapped
    return primal(*args, **kwargs)
  File "/home/oliver/mambaforge/envs/torch2/lib/python3.10/site-packages/taichi/lang/kernel_impl.py", line 1035, in __call__
    return self.launch_kernel(kernel_cpp, *args)
  File "/home/oliver/mambaforge/envs/torch2/lib/python3.10/site-packages/taichi/lang/kernel_impl.py", line 966, in launch_kernel
    raise e from None
  File "/home/oliver/mambaforge/envs/torch2/lib/python3.10/site-packages/taichi/lang/kernel_impl.py", line 959, in launch_kernel
    compiled_kernel_data = prog.compile_kernel(prog.config(), prog.get_device_caps(), t_kernel)
RuntimeError: [frontend_ir.cpp:is_global@850] Assertion failure: var.cast<IndexExpression>()->is_matrix_field() || var.cast<IndexExpression>()->is_ndarray()

```

**Comment by bobcao3** [comment]
That's a bit crazy... let me check

**Comment by bobcao3** [comment]
Should be fixed in #8445 and built in taichi-nightly

**Comment by oliver-batchelor** [comment]
Sweet, thank you!
